tauri-app/src-tauri/src/portapi/client.rs

```diff
@@ -0,0 +1,181 @@
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use futures_util::{SinkExt, StreamExt};
+use tokio::sync::RwLock;
+use tokio::sync::mpsc::{Sender, Receiver, channel};
+use tokio_websockets::{ClientBuilder, Error};
+use http::Uri;
+
+use super::types::*;
+use super::message::*;
+
+/// An internal representation of a Command that
+/// contains the PortAPI message as well as a response
+/// channel that will receive all responses sent from the
+/// server.
+/// 
+/// Users should normally not need to use the Command struct
+/// directly since `PortAPI` already abstracts the creation of
+/// mpsc channels.
+struct Command {
+    msg: Message,
+    response: Sender<Response>
+}
+
+/// The client implementation for PortAPI.
+#[derive(Clone)]
+pub struct PortAPI{
+    dispatch: Sender<Command>,
+}
+
+/// The map type used to store message subscribers.
+type SubscriberMap = RwLock<HashMap<usize, Sender<Response>>>;
+
+/// Connect to PortAPI at the specified URI. 
+/// 
+/// This method will launch a new async thread on the `tauri::async_runtime`
+/// that will handle message to transmit and also multiplex server responses
+/// to the appropriate subscriber.
+pub async fn connect(uri: &str) -> Result<PortAPI, Error> {
+    let parsed = match uri.parse::<Uri>() {
+        Ok(u) => u,
+        Err(e) => {
+            return Err(Error::NoUriConfigured) // TODO(ppacher): fix the return error type.
+        }
+    };
+
+    let (mut client, _) = ClientBuilder::from_uri(parsed).connect().await?;
+    let (tx, mut dispatch) = channel::<Command>(64);
+
+    tauri::async_runtime::spawn(async move {
+        let subscribers: SubscriberMap = RwLock::new(HashMap::new());
+        let next_id = AtomicUsize::new(0);
+
+        loop {
+            tokio::select! {
+                Some(msg) = client.next() => {
+                    match msg {
+                        Err(err) => {
+                            eprintln!("failed to receive frame from websocket: {}", err);
+
+                            return;
+                        },
+                        Ok(msg) => {
+                            let text = unsafe {
+                                std::str::from_utf8_unchecked(msg.as_payload())
+                            };
```

Comment:
Add comment to justify the use of `unsafe` (it's safe in this context since we know it's a text / json string but portmaster unfortunately sends websocket frames as "binary").